### PR TITLE
Remove slow HashSet during miri stack frame creation

### LINKED ISSUE #49237: Regression of compilation times on the latest nightly (2018-03-20)
For [keccak](https://github.com/RustCrypto/sponges) crate I use simple loop unrolling macros:
```Rust
macro_rules! unroll5 {
    ($var:ident, $body:block) => {
        { const $var: usize = 0; $body; }
        { const $var: usize = 1; $body; }
        { const $var: usize = 2; $body; }
        { const $var: usize = 3; $body; }
        { const $var: usize = 4; $body; }
    };
}
```

Which combined generate a lot of unrolled code as a final result. On the latest Nightly it takes a lot more time to compile the crate than previously:
```sh
$ cargo clean; rustup run stable cargo build
   Compiling keccak v0.1.0 (...)
    Finished dev [unoptimized + debuginfo] target(s) in 3.9 secs
cargo clean; rustup run stable cargo check
   Compiling keccak v0.1.0 (...)
    Finished dev [unoptimized + debuginfo] target(s) in 2.30 secs
$ cargo clean; rustup run nightly cargo build
   Compiling keccak v0.1.0 (...)
    Finished dev [unoptimized + debuginfo] target(s) in 18.65 secs
$ cargo clean; rustup run nightly cargo check
   Compiling keccak v0.1.0 (...)
    Finished dev [unoptimized + debuginfo] target(s) in 2.32 secs
```
Here I've used only 10 iterations instead of 24 in the `unroll24` macro, full version takes more than several minutes to compile. Judging by `cargo check` warnings expansion takes approximately the same time and the drastic difference comes from later stages.

EDIT: `RUSTFLAGS="-Z time-passes" rustup run nightly cargo build` produces the following result:
<details>
  <summary>Click to expand</summary>

```
  time: 0.001; rss: 48MB	parsing
  time: 0.000; rss: 50MB	garbage collect incremental cache directory
  time: 0.000; rss: 50MB	recursion limit
  time: 0.000; rss: 50MB	crate injection
  time: 0.000; rss: 50MB	plugin loading
  time: 0.000; rss: 50MB	plugin registration
  time: 0.000; rss: 50MB	background load prev dep-graph
  time: 0.059; rss: 71MB	expansion
  time: 0.000; rss: 71MB	maybe building test harness
  time: 0.001; rss: 71MB	maybe creating a macro crate
  time: 0.003; rss: 71MB	creating allocators
  time: 0.002; rss: 71MB	AST validation
  time: 0.013; rss: 74MB	name resolution
  time: 0.002; rss: 74MB	complete gated feature checking
  time: 0.000; rss: 74MB	blocked while dep-graph loading finishes
  time: 0.016; rss: 80MB	lowering ast -> hir
  time: 0.008; rss: 80MB	early lint checks
  time: 0.021; rss: 83MB	indexing hir
  time: 0.000; rss: 79MB	load query result cache
  time: 0.000; rss: 79MB	looking for entry point
  time: 0.000; rss: 79MB	looking for plugin registrar
  time: 0.001; rss: 79MB	loop checking
  time: 0.000; rss: 81MB	attribute checking
  time: 0.006; rss: 84MB	stability checking
  time: 0.012; rss: 88MB	type collecting
  time: 0.000; rss: 88MB	outlives testing
  time: 0.000; rss: 88MB	impl wf inference
  time: 0.000; rss: 88MB	coherence checking
  time: 0.000; rss: 88MB	variance testing
  time: 0.046; rss: 106MB	wf checking
  time: 0.023; rss: 109MB	item-types checking
  time: 1.518; rss: 121MB	item-bodies checking
  time: 0.031; rss: 122MB	rvalue promotion
  time: 0.014; rss: 122MB	privacy checking
  time: 0.002; rss: 122MB	intrinsic checking
  time: 0.006; rss: 122MB	match checking
  time: 0.084; rss: 118MB	liveness checking
  time: 0.324; rss: 134MB	borrow checking
  time: 0.002; rss: 135MB	MIR borrow checking
  time: 0.000; rss: 135MB	MIR effect checking
  time: 0.003; rss: 135MB	death checking
  time: 0.000; rss: 135MB	unused lib feature checking
  time: 0.039; rss: 138MB	lint checking
  time: 0.000; rss: 138MB	dumping chalk-like clauses
  time: 0.000; rss: 138MB	resolving dependency formats
    time: 0.037; rss: 140MB	write metadata
    time: 14.945; rss: 143MB	translation item collection
    time: 0.000; rss: 143MB	codegen unit partitioning
    time: 0.142; rss: 151MB	translate to LLVM IR
    time: 0.000; rss: 151MB	assert dep graph
    time: 0.000; rss: 152MB	llvm function passes [2lyh15q6cjwzy18c]
    time: 0.000; rss: 152MB	llvm module passes [2lyh15q6cjwzy18c]
    time: 0.002; rss: 159MB	codegen passes [2lyh15q6cjwzy18c]
      time: 0.024; rss: 160MB	persist query result cache
    time: 0.030; rss: 162MB	llvm function passes [30rksvufw6ddw8se]
    time: 0.004; rss: 163MB	llvm module passes [30rksvufw6ddw8se]
      time: 0.011; rss: 161MB	persist dep-graph
    time: 0.034; rss: 161MB	serialize dep graph
  time: 15.160; rss: 161MB	translation
    time: 0.526; rss: 117MB	codegen passes [30rksvufw6ddw8se]
  time: 0.645; rss: 108MB	LLVM passes
  time: 0.000; rss: 108MB	serialize work products
  time: 0.001; rss: 108MB	linking
```
</details>

### PR DESCRIPTION
**Pull request by oli-obk**
fixes #49237

probably has a major impact on #48846

r? @michaelwoerister 

cc @eddyb I know you kept telling me to use vectors instead of hash containers... Now I know why.

### CODE REVIEW

**Review by michaelwoerister — COMMENTED** [review]
Looks good to me!

If this is something that is also called for trivial constants, I would suggest adding a fast path for that case, e.g.:

```rust
if num_locals == 0 {
    vec![]
} else {
    // scan blocks and statements
}
```

**Comment by michaelwoerister** [comment]
r=me with the optimization implemented (if it makes sense).

**Comment by oli-obk** [comment]
Even trivial constants have locals. But you're right, we can add a fast path: Constants and statics don't contain any `StorageDead` or `StorageLive` statements, so we don't need to run this loop at all, irrelevant of the complexity of the constant.

File: src/librustc_mir/interpret/eval_context.rs [hunk 1]
```diff
@@ -383,40 +383,29 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
...
         let num_locals = mir.local_decls.len() - 1;
 
-        let locals = {
-            let annotated_locals = collect_storage_annotations(mir);
-            let mut locals = vec![None; num_locals];
-            for i in 0..num_locals {
-                let local = mir::Local::new(i + 1);
-                if !annotated_locals.contains(&local) {
-                    locals[i] = Some(Value::ByVal(PrimVal::Undef));
+        let mut locals = vec![Some(Value::ByVal(PrimVal::Undef)); num_locals];
```
Comment (eddyb) [inline src/librustc_mir/interpret/eval_context.rs]: Please use `IndexVec` here.

File: src/librustc_mir/interpret/eval_context.rs [hunk 2]
```diff
@@ -383,40 +383,29 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
     ) -> EvalResult<'tcx> {
         ::log_settings::settings().indentation += 1;
 
-        /// Return the set of locals that have a storage annotation anywhere
-        fn collect_storage_annotations<'mir, 'tcx>(mir: &'mir mir::Mir<'tcx>) -> HashSet<mir::Local> {
-            use rustc::mir::StatementKind::*;
-
-            let mut set = HashSet::new();
```
Comment (eddyb) [inline src/librustc_mir/interpret/eval_context.rs]: Would be a good idea to check for `Hash{Map,Set}` used by miri code: if it needs to be a hash map/set, it should be `FxHash{Map,Set}` instead.

**Review by eddyb — CHANGES_REQUESTED** [review]

File: src/librustc_mir/interpret/eval_context.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs]: The reason we didn't use that is theat we don't actually use the return pointer local at all.

File: src/librustc_mir/interpret/eval_context.rs (hunk 2 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs]: Done, found a few in `interpret::memory`

File: src/librustc_mir/interpret/eval_context.rs (hunk 1 above)
Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs]: Fixed, the waste of one space massively outweighs the many `.index() - 1` calls we had.

Comment (michaelwoerister) [reply src/librustc_mir/interpret/eval_context.rs]: Are you sure? Now you have to allocate something on the heap even if it's never used.

Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs]: For truly trivial constants (`const FOO: usize = 42;`) that is true, but even `const FOO: usize = 40 + 2;` will create that heap alloc.

I added a fast path for trivial constants

Comment (michaelwoerister) [reply src/librustc_mir/interpret/eval_context.rs]: :+1:

File: src/librustc_mir/interpret/eval_context.rs [hunk 3]
```diff
@@ -1720,17 +1707,17 @@ impl<'mir, 'tcx> Frame<'mir, 'tcx> {
...
 
     /// Returns the old value of the local
     pub fn storage_dead(&mut self, local: mir::Local) -> EvalResult<'tcx, Option<Value>> {
         trace!("{:?} is now dead", local);
 
-        let old = self.locals[local.index() - 1];
-        self.locals[local.index() - 1] = None;
+        let old = self.locals[local];
+        self.locals[local] = None;
         return Ok(old);
```
Comment (eddyb) [inline src/librustc_mir/interpret/eval_context.rs]: This is just `Ok(self.locals[local].take())`

**Pull request merged**